### ISSUE TITLE
pip install of carbon 0.9.15 = total failure

### ISSUE THREAD

**Issue by jblaine** [graphite-project/carbon#546]
## This simple supported installation ...

```
rm -r /opt/graphite # ON A TEST HOST!
virtualenv /opt/graphite
source /opt/graphite/bin/activate
pip install carbon==0.9.15
```
## .. results in (from the same shell)...

```
bin/carbon-cache.py
Traceback (most recent call last):
  File "/opt/graphite/bin/carbon-cache.py", line 28, in <module>
    from carbon.util import run_twistd_plugin
ImportError: No module named carbon.util
```
## ... because ...

```
(graphite) graphite:etc-metrics# ls lib64/python2.7/site-packages/opt/graphite/lib/
carbon  twisted
(graphite) graphite:etc-metrics#
```

Highlighting the error above in bold: lib64/python2.7/site-packages**/opt/graphite/lib/**

**Comment by jblaine** [comment]
Combined with https://github.com/graphite-project/graphite-web/issues/1508 makes for a totally broken public release to pypi of 0.9.15 :(

**Comment by jblaine** [comment]
The same thing happens with carbon 0.9.14 from pypi.

**Comment by howdoicomputer** [comment]
The version from pypi also does not contain this PR: https://github.com/graphite-project/carbon/pull/53

**Comment by danmac-uk** [comment]
Download and untar the pypi package, edit setup.cfg, comment out the top three lines, re-tar and install.

```
wget https://pypi.python.org/packages/66/ad/7c5fe8471676d45fd80a05c1b7282fb9273b3a3067d2087df8168718349f/carbon-0.9.15.tar.gz
tar xvf carbon-0.9.15.tar.gz
nano carbon-0.9.15/setup.cfg
```
Either clobber or comment with `#` the first three lines, which should be:
```
#[install]
#prefix = /opt/graphite
#install-lib = %(prefix)s/lib
```
Then:
```
tar cvaf carbon-0.9.15-fixed.tar.gz carbon-0.9.15
pip install carbon-0.9.15-fixed.tar.gz
```

I was surprised this package has apparently been broken for over a year - I thought it was something I was doing wrong, but no, the package is broken.

**Comment by deniszh** [comment]
Package is not broken, it's not compatible with latest pip - https://github.com/graphite-project/graphite-web/issues/1508#issuecomment-244200688

**Comment by danmac-uk** [comment]
Thanks for the info :) I think our definitions of broken differ somewhat, but yeah after reading that yarn of hand-wringing I'll leave my comment above and continue with my day :P

**Comment by obfuscurity** [comment]
We understand the issue, and there's nothing we can do it about as it pertains to the 0.9.15 release since bumping it would require a new version anyways. We've already linked to the explanation and workaround in https://github.com/graphite-project/graphite-web/issues/1508#issuecomment-244200688 so I'm going to close this out.

**Comment by danmac-uk** [comment]
May I therefore recommend at least adding some warning to the documentation, even the [latest](https://graphite.readthedocs.io/en/latest/install-virtualenv.html) or to the PyPI page etc as, with respect, the problem here isn't whether the _Graphite team_ understand the issue or not.

After all, I'm just an average guy wanting some pretty graphs to look at, and I fell over this bug, couldn't find anything useful to help, and the path of least resistance was to diagnose and fix the problem myself.

**Comment by obfuscurity** [comment]
@danmac-uk I would love nothing more than to fix the 0.9.15 package so that it's forward compatible with newer pip, or at least, update the 0.9.15 documentation to explicitly mention the workaround. I've explained [here](https://github.com/graphite-project/graphite-web/issues/1508#issuecomment-244383013) why it's not as easy as it sounds. If someone else has a practical suggestion I'm all ears.

**Comment by danmac-uk** [comment]
I'm not intimately familiar with PyPI packaging but at the very least I would add something to the latest docs which I assume would be uncomplicated.

I'm sure I'm not fully aware of the situation but if "blow away three lines from setup.cfg and release 0.9.16" isn't an option then yeah it does sound like a conundrum :(

**Comment by obfuscurity** [comment]
No worries, the fact that you're not familiar with PyPI packaging is a sign of mental wellness. :wink:

Long story short, Graphite's docs are tightly coupled to the versions they represented. IOW, updating the 0.9.15 docs would require you to ship a new version (0.9.16). It's a clusterfuck.

**Comment by danmac-uk** [comment]
Oh, yeah I got that already. Sorry I was just pointing out that a note on the _latest_ docs would have helped me, or a 0.9.16 version with the three lines blown away.

I can understand each package version being immutable upon release but the docs too, yeah hmmm .... :/

**Comment by obfuscurity** [comment]
To repair backwards compatibility due to pip changes? Virtually zero. If there are security-related reasons then we'd certainly consider it.

Of course I'm just one maintainer/developer. I don't want to speak for the entire project. If there's enough interest in doing such a thing, by all means don't let me hold anyone back.

**Comment by deniszh** [comment]
IIRC it was decided to release 1.0.0 based on master branch, and not release any 0.9.x anymore.
Maybe it's worth to release 0.9.16 as a final version of 0.9.x together with 1.0.0...

**Comment by howdoicomputer** [comment]
I was hoping more to pick up the delta between what is published on PyPi and what exists on the 0.9.X branch. However, if a 0.10.0 is around the corner (or a 1.0.0) then I'm fairly content.

**Comment by deniszh** [comment]
We decided to release 0.9.16 together with 1.0.0 btw.
